extensions/machine-learning-services/src/packageManagement/sqlRPackageManageProvider.ts

```diff
@@ -0,0 +1,154 @@
+		private _processService: ProcessService,
+		private _config: Config) {
+		super(apiWrapper);
+		this._rExecutable = this._config.rExecutable;
+	}
+
+	/**
+	 * Returns provider Id
+	 */
+	public get providerId(): string {
+		return SqlRPackageManageProvider.ProviderId;
+	}
+
+	/**
+	 * Returns package target
+	 */
+	public get packageTarget(): nbExtensionApis.IPackageTarget {
+		return { location: 'SQL', packageType: 'R' };
+	}
+
+	/**
+	 * Returns list of packages
+	 */
+	public async listPackages(): Promise<nbExtensionApis.IPackageDetails[]> {
+		let packages = await this._queryRunner.getRPackages(await this.getCurrentConnection());
+		if (packages) {
+			packages = packages.sort((a, b) => a.name.localeCompare(b.name));
+		} else {
+			packages = [];
+		}
+		return packages;
+	}
+
+	/**
+	 * Installs given packages
+	 * @param packages Packages to install
+	 * @param useMinVersion minimum version
+	 */
+	async installPackages(packages: nbExtensionApis.IPackageDetails[], useMinVersion: boolean): Promise<void> {
+		if (packages) {
+
+			// TODO: install package as parallel
+			for (let index = 0; index < packages.length; index++) {
+				const element = packages[index];
+				await this.executeScripts(installMode, element);
+			}
+		}
+		//TODO: use useMinVersion
+		console.log(useMinVersion);
+	}
+
+	/**
+	 * Execute a script to install or uninstall a python package inside current SQL Server connection
+	 * @param packageDetails Packages to install or uninstall
+	 * @param scriptMode can be 'install' or 'uninstall'
+	 */
+	private async executeScripts(scriptMode: string, packageDetails: nbExtensionApis.IPackageDetails): Promise<void> {
+		let connection = await this.getCurrentConnection();
+		let credentials = await this._apiWrapper.getCredentials(connection.connectionId);
+
+		if (connection) {
+			let database = connection.databaseName ? `, database="${connection.databaseName}"` : '';
+			let pythonConnectionParts = `server="${connection.serverName}", uid="${connection.userName}", pwd="${credentials[azdata.ConnectionOptionSpecialType.password]}"${database}`;
+			let rCommandScript = scriptMode === installMode ? 'sql_install.packages' : 'sql_remove.packages';
+
+			let scripts: string[] = [
+				'formals(quit)$save <- formals(q)$save <- "no"',
+				'library(sqlmlutils)',
+				`connection <- connectionInfo(${pythonConnectionParts})`,
+				`pkgs <- c("${packageDetails.name}")`,
+				`${rCommandScript}(connectionString = connection, pkgs, scope = "PUBLIC")`,
+				'q()'
+			];
+			await this._processService.execScripts(`${this._rExecutable}`, scripts, ['--vanilla'], this._outputChannel);
+		}
+	}
+
+	/**
+	 * Uninstalls given packages
+	 * @param packages Packages to uninstall
+	 */
+	async uninstallPackages(packages: nbExtensionApis.IPackageDetails[]): Promise<void> {
+		for (let index = 0; index < packages.length; index++) {
+			const element = packages[index];
+			await this.executeScripts(uninstallMode, element);
```

Comment:
What's the behavior if one fails?  Finish uninstalling the rest, or fail the entire batch?

Reply:
I think it should try to uninstall as many as it can. will change to promise all

---
In reply to: [366620248](https://github.com/microsoft/azuredatastudio/pull/8870#discussion_r366620248) [](ancestors = 366620248)